### PR TITLE
(CDAP-17611) Added OAuth function support

### PR DESCRIPTION
**Pull request by chtyim**

### CODE REVIEW

**Review by CuriousVini — APPROVED** [review]

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceOutputFormatProvider.java [hunk 1]
```diff
@@ -34,17 +36,25 @@
...
       .put(SalesforceSinkConstants.CONFIG_SOBJECT, config.getSObject())
       .put(SalesforceSinkConstants.CONFIG_OPERATION, config.getOperation())
       .put(SalesforceSinkConstants.CONFIG_ERROR_HANDLING, config.getErrorHandling().getValue())
       .put(SalesforceSinkConstants.CONFIG_MAX_BYTES_PER_BATCH, config.getMaxBytesPerBatch().toString())
       .put(SalesforceSinkConstants.CONFIG_MAX_RECORDS_PER_BATCH, config.getMaxRecordsPerBatch().toString());
 
+    OAuthInfo oAuthInfo = config.getOAuthInfo();
+    if (oAuthInfo != null) {
+      configBuilder.put(SalesforceConstants.CONFIG_OAUTH_TOKEN, oAuthInfo.getAccessToken());
+      configBuilder.put(SalesforceConstants.CONFIG_OAUTH_INSTANCE_URL, oAuthInfo.getInstanceURL());
```
Comment (tivv) [inline src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceOutputFormatProvider.java]: nit: use same style (chained or not of builder calls in both `if` and `else`

**Review by tivv — APPROVED** [review]

File: src/main/java/io/cdap/plugin/salesforce/plugin/BaseSalesforceConfig.java [hunk 2]
```diff
@@ -32,65 +32,95 @@
  */
 public class BaseSalesforceConfig extends ReferencePluginConfig {
 
+  @Name(SalesforceConstants.PROPERTY_OAUTH_INFO)
+  @Description("OAuth information for connecting to Salesforce. " +
+    "It is expected to be an json string containing two properties, \"accessToken\" and \"instanceURL\", " +
+    "which carry the OAuth access token and the URL to connect to respectively. " +
+    "Use the ${oauth(provider, credentialId)} macro function for getting acquiring OAuth information dynamically. ")
```
Comment (supriyapatil-google) [inline src/main/java/io/cdap/plugin/salesforce/plugin/BaseSalesforceConfig.java]: remove getting.

File: src/main/java/io/cdap/plugin/salesforce/plugin/OAuthInfo.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.salesforce.plugin;
+
+/**
+ * Class to carry OAuth information returned by the {@code ${oauth}} macro function.
+ */
+public class OAuthInfo {
```
Comment (supriyapatil-google) [inline src/main/java/io/cdap/plugin/salesforce/plugin/OAuthInfo.java]: final

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceOutputFormatProvider.java [hunk 4]
```diff
@@ -34,17 +36,25 @@
    */
   public SalesforceOutputFormatProvider(SalesforceSinkConfig config) {
     ImmutableMap.Builder<String, String> configBuilder = new ImmutableMap.Builder<String, String>()
-      .put(SalesforceConstants.CONFIG_USERNAME, config.getUsername())
-      .put(SalesforceConstants.CONFIG_PASSWORD, config.getPassword())
-      .put(SalesforceConstants.CONFIG_CONSUMER_KEY, config.getConsumerKey())
-      .put(SalesforceConstants.CONFIG_CONSUMER_SECRET, config.getConsumerSecret())
-      .put(SalesforceConstants.CONFIG_LOGIN_URL, config.getLoginUrl())
       .put(SalesforceSinkConstants.CONFIG_SOBJECT, config.getSObject())
```
Comment (supriyapatil-google) [inline src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceOutputFormatProvider.java:39]: SUBJECT

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceOutputFormatProvider.java (hunk 1 above)
Comment (chtyim) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceOutputFormatProvider.java]: fixed

**Review by supriyapatil-google — APPROVED** [review]
Thanks!

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceOutputFormatProvider.java (hunk 4 above)
Comment (chtyim) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceOutputFormatProvider.java:39]: I think SOJBECT stands for the Salesforce object.

File: src/main/java/io/cdap/plugin/salesforce/plugin/BaseSalesforceConfig.java (hunk 2 above)
Comment (chtyim) [reply src/main/java/io/cdap/plugin/salesforce/plugin/BaseSalesforceConfig.java]: removed

File: src/main/java/io/cdap/plugin/salesforce/plugin/OAuthInfo.java (hunk 3 above)
Comment (chtyim) [reply src/main/java/io/cdap/plugin/salesforce/plugin/OAuthInfo.java]: changed

**Pull request merged**